tracking-migration/src/main/java/io/customer/tracking/migration/util/JsonAdapter.kt

```diff
@@ -11,4 +22,101 @@ class JsonAdapter {
+                    event = eventJson.requireField("name") {
+                        val value = stringOrNull(it)
+                        remove(it)
+                        return@requireField value
+                    },
+                    type = eventJson.requireField("type") {
+                        val value = stringOrNull(it)
+                        remove(it)
+                        return@requireField value
+                    },
+                    properties = eventJson
+                )
+            }
+
+            QueueTaskType.RegisterDeviceToken -> {
+                val deviceJson = taskJson.requireField("device") { jsonObjectOrNull(it) }
+
+                MigrationTask.RegisterDeviceToken(
+                    timestamp = timestamp,
+                    identifier = taskJson.requireField("profileIdentified") { stringOrNull(it) },
+                    token = deviceJson.requireField("id") { stringOrNull(it) },
+                    platform = deviceJson.requireField("platform") { stringOrNull(it) },
+                    lastUsed = deviceJson.longOrNull("last_used") ?: deviceJson.longOrNull("lastUsed") ?: timestamp,
+                    attributes = deviceJson.jsonObjectOrNull("attributes") ?: JSONObject()
+                )
+            }
+
+            QueueTaskType.DeletePushToken -> {
+                MigrationTask.DeletePushToken(
+                    timestamp = timestamp,
+                    identifier = taskJson.requireField("profileIdentified") { stringOrNull(it) },
+                    token = taskJson.requireField("deviceToken") { stringOrNull(it) }
+                )
+            }
+
+            QueueTaskType.TrackPushMetric -> {
+                MigrationTask.TrackPushMetric(
+                    timestamp = timestamp,
+                    deliveryId = taskJson.requireField("delivery_id") { stringOrNull(it) },
+                    deviceToken = taskJson.requireField("device_id") { stringOrNull(it) },
+                    event = taskJson.requireField("event") { stringOrNull(it) }
+                )
+            }
+
+            QueueTaskType.TrackDeliveryEvent -> {
+                val payloadJson = taskJson.requireField("payload") { jsonObjectOrNull(it) }
+
+                MigrationTask.TrackDeliveryEvent(
+                    timestamp = payloadJson.longOrNull("timestamp") ?: timestamp,
+                    deliveryType = taskJson.requireField("type") { stringOrNull(it) },
+                    deliveryId = payloadJson.requireField("delivery_id") { stringOrNull(it) },
+                    event = payloadJson.requireField("event") { stringOrNull(it) },
+                    metadata = payloadJson.jsonObjectOrNull("metadata") ?: JSONObject()
+                )
+            }
+        }
+
+        return Result.success(result)
+    }
+}
+
+private fun <T> JSONObject?.requireField(key: String, getter: JSONObject.(String) -> T?): T {
```

Comment:
think this can be simplified, 

```
private inline fun <reified T> JSONObject.requireField(key: String): T {
    return when (T::class) {
        String::class -> this.stringOrNull(key)
        else -> this.opt(key)
    } as? T ?: throw IllegalArgumentException("Required key '$key' is missing or has wrong type in $this")
}
```

now we can just use, e.g for json object
```
val payloadJson = taskJson.requireField<JSONObject>("payload")
```
for string
```
deliveryType = taskJson.requireField("type"),
```

we can check for type in runtime

Reply:
Good call. The only concern I have is it will suppress any type that the callers tries to parse and is not currently supported as the whole code is in `runCatching`. I have updated it to more generic implementation and added docs to reduce the possibility of this, but we should still be careful when using this so that we don't keep suppressing it unknowingly.